cmd/gateway-gcs.go

```diff
@@ -561,266 +570,190 @@ func (l *gcsGateway) DeleteObject(bucket string, object string) error {
 	uploadID = mustGetUUID()
-	uploadID = strings.Replace(uploadID, "-", "", -1)
 
 	// generate name for part zero
-	partZeroKey := toGCSMultipartKey(key, uploadID, 0)
-
+	meta := gcsMultipartMetaName(uploadID)
 	// we are writing a 0 sized object to hold the metadata
-	w := l.client.Bucket(bucket).Object(partZeroKey).NewWriter(l.ctx)
+	w := l.client.Bucket(bucket).Object(meta).NewWriter(l.ctx)
 	w.ContentType = metadata["content-type"]
 	w.ContentEncoding = metadata["content-encoding"]
 	w.Metadata = metadata
+
+	content, err := json.Marshal(gcsMultipartMetaV1{gcsMultipartMetaCurrentVersion, bucket, key})
+	if err != nil {
+		return "", gcsToObjectError(traceError(err), bucket, key)
+	}
+	if _, err = w.Write(content); err != nil {
+		return "", gcsToObjectError(traceError(err), bucket, key)
+	}
 	if err = w.Close(); err != nil {
 		return "", gcsToObjectError(traceError(err), bucket, key)
 	}
 
 	return uploadID, nil
 }
 
+// ListMultipartUploads - lists all multipart uploads.
+func (l *gcsGateway) ListMultipartUploads(bucket string, prefix string, keyMarker string, uploadIDMarker string, delimiter string, maxUploads int) (ListMultipartsInfo, error) {
+	return ListMultipartsInfo{
+		KeyMarker:      keyMarker,
+		UploadIDMarker: uploadIDMarker,
+		MaxUploads:     maxUploads,
+		Prefix:         prefix,
+		Delimiter:      delimiter,
+	}, nil
+}
+
 // CopyObjectPart - copy part of object to other bucket and object
 func (l *gcsGateway) CopyObjectPart(srcBucket string, srcObject string, destBucket string, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
 	return PartInfo{}, traceError(NotSupported{})
 }
 
 // PutObjectPart puts a part of object in bucket
 func (l *gcsGateway) PutObjectPart(bucket string, key string, uploadID string, partID int, size int64, data io.Reader, md5Hex string, sha256sum string) (PartInfo, error) {
-	multipartKey := toGCSMultipartKey(key, uploadID, partID)
+	meta := gcsMultipartMetaName(uploadID)
+	object := l.client.Bucket(bucket).Object(meta)
 
-	info, err := l.PutObject(bucket, multipartKey, size, data, map[string]string{}, sha256sum)
-	return PartInfo{
-		PartNumber:   partID,
-		LastModified: info.ModTime,
-		ETag:         info.ETag,
-		Size:         info.Size,
-	}, err
-}
+	_, err := object.Attrs(l.ctx)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-// ListObjectParts returns all object parts for specified object in specified bucket
-func (l *gcsGateway) ListObjectParts(bucket string, key string, uploadID string, partNumberMarker int, maxParts int) (ListPartsInfo, error) {
-	delimiter := slashSeparator
-	prefix := pathJoin(ZZZZMinioPrefix, fmt.Sprintf("multipart-%s-%s", escape(key), uploadID))
+	var sha256Writer hash.Hash
 
-	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: delimiter, Prefix: prefix, Versions: false})
+	// Generate random ETag.
+	etag := getMD5Hash([]byte(mustGetUUID()))
```

Comment:
Why is `mustGetUUID()` alone not sufficient if the etag just needs to be random?